cpp/src/io/parquet/writer_impl.cu

```diff
@@ -384,15 +384,18 @@ writer::impl::impl(std::unique_ptr<data_sink> sink, writer_options const &option
   stats_granularity_(options.stats_granularity),
   out_sink_(std::move(sink)){}
 
-void writer::impl::write(table_view const &table, const table_metadata *metadata, cudaStream_t stream) {
+std::unique_ptr<std::vector<uint8_t>> writer::impl::write(
+                         table_view const &table, const table_metadata *metadata,
+                         bool return_filemetadata, const std::string& metadata_out_file_path,
+                         cudaStream_t stream) {    
```

Comment:
Do you know if there are other APIs in cuDF that optionally return something based on a `boolean` parameter?

Reply:
I would guess its not a very common situation.

Reply:
That's a good observation, @codereport . Is it possible (and efficient) to just make a separate function to get the metadata?

Reply:
Yea, it makes me uncomfortable have a implicitly "optional" return value that is determined by a boolean parameter.

Reply:
Part of the metadata contains statistics from the write operation. Returning it in a separate call would imply the writer maintaining state (doesn't really work for the python side), or huge perf hit (second dummy write operation).

Reply:
Also, I don't understand how it's `implicitly optional` if it's explicitly dependent on an input parameter. To be clear, the boolean parameter allows for an optimization (as this is mainly needed for dask). We could eliminate this parameter and either always return the metadata or never return the metadata (what we have now), at the cost of injecting some mediocrity in the parquet writer performance for satisfying a demand for conformity.

Reply:
IMO, other than duplicating the API, the only clean way to make returning the statistics a separate call would be to eliminate the C-style IO APIs in functions.hpp along with the various read/write args structs (which might not be a bad thing, since it's largely a legacy of the pre-cython days afaik) and have the cython layer call the reader/writer classes directly (which would allow moving the intermediate writer state into the class). That should probably be a separate issue, though, since it involves a significant rewrite of the python/cython layer for all io readers & writers (with the benefit of eliminating the APIs in io/functions.hpp).

Reply:
Got it. Thanks for the detail, Olivier.

Reply:
I agree that moving away from C-style APIs is a good thing. And if this is mainly an artifact of such APIs in `functions.hpp` than I agree, separating it into a separate issue is a probably the way to go (and I am fine with the change as is). 

I will try to explain how this feels `implicitly optional` though. In C++17, we got `std::optional` for when you invoke a function that might not return a valid result. The classic example is `find_if`. Our `find_if` will current return `std::end()` of the data structure if it is unable to find an element that matches a predicate (in other languages that had `optional` or `maybe` from the beginning, the return those). Functions / algorithms like this are explicitly optional in that you don't know going in whether or not you will get a value back (or nothing back), which is why you get back an "optional" result. 

However, for this function - going into it we **know** whether or not we will get back a value - because we are specifying it in the interface. If we know, we should be separating the APIs to avoid the unnecessary complexity. Maybe "implict" is the wrong word and "unnecessarily optional" is a better way to describe it. To me - it just feels like the wrong API design. 

However, as I said at the beginning - I agree that this probably isn't the PR to address and the changes you have are appropriate given the foundation you are working on. This was just to try and explain why it "made me uncomfortable".

Thanks for explaining the background and details.